### PR TITLE
Add password reset

### PR DESCRIPTION
**Pull request by thanhconguet214**

### CODE REVIEW

**Comment by thanhconguet214** [comment]
ready

File: app/assets/javascripts/password_resets.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (huuphuc93) [inline app/assets/javascripts/password_resets.coffee]: remove this file file

File: app/assets/stylesheets/password_resets.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the PasswordResets controller here.
```
Comment (huuphuc93) [inline app/assets/stylesheets/password_resets.scss]: remove this file file

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      redirect_to @user
+    else
+      render :edit
+    end
+  end
+
+  private
+
+  def user_params
+      params.require(:user).permit :password, :password_confirmation
```
Comment (huuphuc93) [inline app/controllers/password_resets_controller.rb]: indent 2 spaces instead of 4

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def get_user
+      @user = User.find_by email: params[:email]
```
Comment (huuphuc93) [inline app/controllers/password_resets_controller.rb]: indent

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def valid_user
+    unless (@user && @user.activated? && @user.authenticated?(:reset, params[:id]))
```
Comment (huuphuc93) [inline app/controllers/password_resets_controller.rb]: use return if insted of unless

File: app/helpers/password_resets_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
```
Comment (huuphuc93) [inline app/helpers/password_resets_helper.rb]: remove this file

File: app/views/user_mailer/account_activation.html.erb [hunk 7]
```diff
@@ -1,6 +1,5 @@
 <h1><%= t "user_mailer.account_activation.sample" %></h1>
 <p><%= t "user_mailer.account_activation.hi" %><%= @user.name %>,</p>
 <p><%= t "user_mailer.account_activation.welcome" %></p>
-<%= link_to t("user_mailer.account_activation.activate"), edit_account_activation_url(@user.activation_token,
-  email: @user.email) %>
+<%= link_to t("user_mailer.account_activation.activate"), edit_account_activation_url(@user.activation_token, email: @user.email) %>
```
Comment (huuphuc93) [inline app/views/user_mailer/account_activation.html.erb]: eof

**Comment by huuphuc93** [comment]
commented

**Comment by thanhconguet214** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user,   only: %i(edit update)
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: Please remove unnecessary space

File: app/controllers/password_resets_controller.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  before_action :valid_user, only: %i(edit update)
+  before_action :check_expiration, only: %i(edit update)
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: As the actions are same, you can combine lines 2, 3 and 4 into one

File: app/controllers/password_resets_controller.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+  def new; end
+
+  def create
+    @user = User.find_by email: params[:password_reset][:email].downcase 
+    if @user
+      @user.create_reset_digest
+      @user.send_password_reset_email
+      flash[:info] = t("controller.password_resets.email.sent")
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: Please remove optional ()

File: app/controllers/password_resets_controller.rb [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+      redirect_to root_url
+    else
+      flash.now[:danger] = t("controller.password_resets.email_not_found")
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: Please remove ()

File: app/controllers/password_resets_controller.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    end
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def get_user
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: Bad naming. You shouldn't name a method in ruby like as set/get. You can name it like as load_user

File: app/controllers/password_resets_controller.rb [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    @user = User.find_by email: params[:email]
+  end
+
+  def valid_user
+    return if (@user && @user.activated? && @user.authenticated?(:reset, params[:id]))
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: Remove redundant ()

File: app/controllers/password_resets_controller.rb [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+      redirect_to root_url
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: 1. Indentation
2. Before redirection add a flash message to inform the requesting user

File: app/controllers/password_resets_controller.rb [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  def check_expiration
+    if @user.password_reset_expired?
+      flash[:danger] = t "controller.password_resets.check_expiration"
+      redirect_to new_password_reset_url
+    end
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]:
```ruby
return unless @user.password_reset_expired?
flash[:danger] = t "controller.password_resets.check_expiration"
redirect_to new_password_reset_url
```

File: app/views/user_mailer/password_reset.text.erb [hunk 16]
```diff
@@ -1,2 +1,4 @@
-<%= t "user_mailer.account_activation.p1" %>
-<%= @greeting %>, <%= t "user_mailer.account_activation.content" %>
+<%= t "user_mailers.password_reset.link_reset" %>
+<%= edit_password_reset_url(@user.reset_token, email: @user.email) %>
```
Comment (mdmahmudurrahman) [inline app/views/user_mailer/password_reset.text.erb]: Please remove ()

File: config/locales/en.yml [hunk 17]
```diff
@@ -67,8 +89,10 @@ en:
...
-
-  mailers:
-    hi: "Hi"
-    account: "Account activation"
+    password_reset:
+        title: "Reset Password"
+        link_reset: "To reset your password click the link below:"
+        expire: "This link will expire in two hours."
+        message: "If you did not request your password to be reset,
+          please ignore this email and your password will stay as it is."
```
Comment (mdmahmudurrahman) [inline config/locales/en.yml]: Add a line break at the end of a file

File: config/locales/en.yml [hunk 18]
```diff
@@ -67,8 +89,10 @@ en:
       active: "Activate"
       p1: "Password Reset"
       content: "find me in app/views/user_mailer/password_reset.html.erb"
-
-  mailers:
-    hi: "Hi"
-    account: "Account activation"
+    password_reset:
+        title: "Reset Password"
```
Comment (mdmahmudurrahman) [inline config/locales/en.yml]: Indentation

File: config/routes.rb [hunk 19]
```diff
@@ -9,5 +11,6 @@
   post "/login", to: "sessions#create"
   delete "/logout", to: "sessions#destroy"
   resources :users
-  resources :account_activations, only: :edit
+  resources :account_activations, only: %i(edit)
```
Comment (mdmahmudurrahman) [inline config/routes.rb]: For one action, no need to create an array
`resources :account_activations, only: :edit`

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by thanhconguet214** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 20]
```diff
@@ -0,0 +1,55 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, :valid_user, :check_expiration only: %i(edit update)
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: Do you have any method named get_user now ?

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by thanhconguet214** [comment]
ready

**Pull request merged**